File: frontend/src/lib/api/assignment.ts
/**
 * API Client untuk modul Assignment (Lead_SA)
 *
 * Fungsi-fungsi untuk:
 * - getPendingAssignment: Ambil daftar proyek menunggu assignment
 * - getAvailableSA: Ambil daftar SA tersedia + workload
 * - assignProject: Assign SA ke proyek
 */

import { fetcher, apiRequest } from "../fetcher";

// === Tipe Data ===

/** Proyek yang menunggu assignment */
export interface PendingProject {
  id_project: string;
  project_name: string;
  customer_name: string;
  bant_score: number;
  use_case_tags: string[];
  status: string;
  target_submit: string;
  gdrive_folder_id: string | null;
  created_at: string;
  sales_pic: {
    id: string;
    name: string;
    email: string;
  };
}

/** SA yang tersedia untuk assignment */
export interface AvailableSA {
  id: string;
  name: string;
  email: string;
  avatar_url: string | null;
  active_project_count: number;
}

/** Response setelah assignment berhasil */
export interface AssignmentResponse {
  id_project: string;
  assigned_sa: string;
  assigned_at: string;
  status: string;
}

/** Ringkasan proyek per status (untuk overview) */
export interface ProjectStatusSummary {
  status: string;
  count: number;
}

// === API Functions ===

/**
 * Ambil daftar proyek berstatus "Pending Assignment"
 * Digunakan di dashboard Lead_SA untuk antrian assignment
 */
export async function getPendingAssignment(): Promise<PendingProject[]> {
  return fetcher<PendingProject[]>("/projects/pending-assignment");
}

/**
 * Ambil daftar SA yang tersedia beserta jumlah proyek aktif
 * Digunakan di modal assignment untuk memilih SA
 */
export async function getAvailableSA(): Promise<AvailableSA[]> {
  return fetcher<AvailableSA[]>("/sa/available");
}

/**
 * Assign SA ke proyek tertentu
 * Hanya bisa dilakukan oleh Lead_SA
 *
 * @param projectId - ID proyek yang akan di-assign
 * @param saId - ID SA yang ditugaskan
 * @returns Data assignment yang berhasil
 */
export async function assignProject(
  projectId: string,
  saId: string
): Promise<AssignmentResponse> {
  return apiRequest<AssignmentResponse>(`/projects/${projectId}/assign`, {
    method: "POST",
    body: { sa_id: saId },
  });
}

File: backend/app/schemas/assignment.py
"""
Schema Pydantic untuk endpoint assignment SA ke proyek.
Digunakan untuk validasi request/response pada fitur penugasan oleh Lead_SA.
"""

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class AssignRequest(BaseModel):
    """Request body untuk assign SA ke proyek."""

    sa_id: uuid.UUID = Field(
        ...,
        description="UUID dari SA yang akan ditugaskan ke proyek",
    )


class SAAvailability(BaseModel):
    """Data ketersediaan SA beserta jumlah proyek aktif."""

    id: uuid.UUID = Field(..., description="UUID SA")
    name: str = Field(..., description="Nama lengkap SA")
    email: str = Field(..., description="Email SA")
    active_project_count: int = Field(
        ...,
        description="Jumlah proyek aktif yang sedang ditangani SA",
    )


class ProjectBrief(BaseModel):
    """Ringkasan data proyek untuk response assignment."""

    id_project: str
    project_name: str
    customer_name: str
    status: str
    bant_score: Optional[int] = None
    use_case_tags: Optional[list] = None
    assigned_sa: Optional[uuid.UUID] = None
    assigned_at: Optional[datetime] = None
    target_submit: date
    gdrive_folder_id: Optional[str] = None


class AssignmentResponse(BaseModel):
    """Response setelah assignment berhasil."""

    project: ProjectBrief = Field(..., description="Data proyek setelah di-assign")
    assigned_sa_name: str = Field(..., description="Nama SA yang ditugaskan")
    notification_sent: bool = Field(
        ...,
        description="Apakah notifikasi berhasil dikirim ke SA",
    )
    folder_provisioning_triggered: bool = Field(
        ...,
        description="Apakah folder provisioning sudah di-trigger",
    )
